ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -259,6 +261,11 @@ data ShakeExtras = ShakeExtras
     ,publishedDiagnostics :: STM.Map NormalizedUri [Diagnostic]
     -- ^ This represents the set of diagnostics that we have published.
     -- Due to debouncing not every change might get published.
+    ,semanticTokensCache:: STM.Map NormalizedUri SemanticTokens
```

Comment:
I think these could probably be variables local to the plugin? I think we only need to put thing in here if they're really shared across multiple rules etc. 

So the function that creates the plugin would run in `IO` or similar and create the TVars, which then get passed into the handlers.

Reply:
Agree that putting it in shakeExtras is not optimal, but I am also not sure about local state to plugin too.
Since no other plugins have their own plugin state. It would mean we need to change `idePlugins :: Recorder (WithPriority Log) -> IdePlugins IdeState` to `idePlugins :: Recorder (WithPriority Log) -> IO (IdePlugins IdeState)`.
And open a door to introduce state to plugins breaking the current centralized state management in the hls build system. 🤔

Reply:
I don't think plugin-local state is inherently problematic. But I admit that it's not totally clear to me what stuff _should_ be in `ShakeExtras` :thinking: thoughts @fendor ?

Reply:
added a comment to the shakeExtra stating it might not be ideal to store the varaibles there, but we do not find a better place to store it yet.

Reply:
I do not have any specific thoughts, if the variable could be easily plugin local, I think I would prefer that.